File: modules/my-module/android/src/main/java/expo/modules/mymodule/MyModule.kt
package expo.modules.mymodule

import android.content.Context
import android.content.SharedPreferences
import android.content.res.Configuration
import androidx.core.os.bundleOf
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class MyModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("MyModule")

    Events("onChangeTheme")

    Function("setTheme") { theme: String ->
      getPreferences().edit().putString("theme", theme).apply()
      val nextTheme = if (theme == "system") getNativeTheme() else theme
      this@MyModule.sendEvent("onChangeTheme", bundleOf("theme" to nextTheme))
    }

    Function("getTheme") {
      val savedTheme = getPreferences().getString("theme", null)
      return@Function savedTheme ?: getNativeTheme()
    }

    Function("getNativeTheme") {
      return@Function getNativeTheme()
    }

    Function("getResolvedTheme") {
      val savedTheme = getPreferences().getString("theme", null)
      return@Function when {
        savedTheme == null -> getNativeTheme()
        savedTheme == "system" -> getNativeTheme()
        else -> savedTheme
      }
    }
  }

  private val context
  get() = requireNotNull(appContext.reactContext)

  private fun getPreferences(): SharedPreferences {
    return context.getSharedPreferences(context.packageName + ".settings", Context.MODE_PRIVATE)
  }

  private fun getNativeTheme(): String {
    val nightModeFlags = context.resources.configuration.uiMode and Configuration.UI_MODE_NIGHT_MASK
    return when (nightModeFlags) {
      Configuration.UI_MODE_NIGHT_YES -> "dark"
      Configuration.UI_MODE_NIGHT_NO -> "light"
      else -> "light"
    }
  }
}

File: modules/my-module/ios/MyModule.swift
import ExpoModulesCore
import UIKit

public class MyModule: Module {
  public func definition() -> ModuleDefinition {
    Name("MyModule")

    Events("onChangeTheme")

    Function("setTheme") { (theme: String) -> Void in
      UserDefaults.standard.set(theme, forKey:"theme")
      let nextTheme = (theme == "system") ? self.getNativeTheme() : theme
      sendEvent("onChangeTheme", [
        "theme": nextTheme
      ])
    }

    Function("getTheme") { () -> String in
      if let savedTheme = UserDefaults.standard.string(forKey: "theme") {
        return savedTheme
      }
      return self.getNativeTheme()
    }

    Function("getNativeTheme") { () -> String in
      return self.getNativeTheme()
    }

    Function("getResolvedTheme") { () -> String in
      if let savedTheme = UserDefaults.standard.string(forKey: "theme") {
        if savedTheme == "system" {
          return self.getNativeTheme()
        }
        return savedTheme
      }
      return self.getNativeTheme()
    }
  }

  private func getNativeTheme() -> String {
      let userInterfaceStyle = UIScreen.main.traitCollection.userInterfaceStyle
      switch userInterfaceStyle {
        case .dark:
            return "dark"
        case .light:
            return "light"
        default:
            return "light"
      }
  }
}
